docs/proposals/subscriber-fan-in-fan-out.md

```diff
@@ -0,0 +1,189 @@
+performs all the work. As mentioned this means no more messages can be
+read from the channel until this unit of work is complete. A change to
+how the messages are handled could alleviate any potential issue. The
+reader would change to be the work engine which would start a read loop
+for each of the topics when the broker started and distribute incoming
+messages to any number of subscribers async:
+As long as the principal that subscribers shouldn't depend on each other
+is maintained there shouldn't be issues with this approach.
+```go 
+for{
+
+   select{
+       case msg := <-provisionChan:
+           for _,s := range e.subscribers[ProvisionTopic]{
+                go s(msg)              
+           }
+       case <-stop:
+       break
+   }
+}
+
+
+```
+
+The stop channel here would be tied to os signals
+[example](https://gist.github.com/reiki4040/be3705f307d3cd136e85).
+
+In this case, we would still have a single channel per topic, but as we
+hand off the message immediately and don't block waiting for the
+subscribers to finish their work, there is little to no chance for the
+channel to become full and so block the running jobs.
+
+#### Add work engine start method
+
+The work engine will now be the one reading messages from the channel
+and passing it on to the Observers. To do this I propose we add a  public
+``Start(stop <-chan struct{})`` method, that would be called in its own
+go routine during ASB start up. I also propose adding a signal channel
+that will send a message to anything running  in the background
+(could also use context.Context). This would signal for the background
+loops etc to stop.
+This start method would start a loop for all of the topics as shown
+above and close them down once a message was received.
+
+#### 2) Create a channel per job
+
+Remove the single channel per topic and replace it with a channel per
+job. The work engine would be updated to create a channel per Job
+when starting a new job. The work engine would be responsible for
+the lifecycle of these channels. As above, the subscribers would be
+registered in a map and called in a non-blocking way when a message was received.
+While more complex, this method would allow us to leverage more control
+(use a buffer to control how much could be sent while still
+not impacting on any other Jobs that running).
+
+Sudo Code:
+
+```go 
+
+StartNewJob(token string, work Work, topic WorkTopic){
+  a.jobs[token] = make(chan JobMsg)
+  go func(){
+      //make sure we always close the channel  
+      defer close(e.jobs[token])
+      // start reading from the channel 
+      go func(){
+      // will auto stop once the channel is closed
+        for msg := range a.jobs[token]{
+            for _, s := range a.subscribers[topic]{
+               // Note we could choose to run these sync but again if there was a slow
+               // subscriber it would block receiving on the channel.
+               go s(msg)
```

Comment:
Same comment WRT `Notify` and what we should do about errors.

Reply:
see above